nilearn/regions/hierarchical_kmeans_clustering.py

```diff
@@ -0,0 +1,276 @@
+    parcels, and Kmeans a second time on each parcel. s
+
+    Parameters
+    ----------
+    X: ndarray (n_samples, n_features)
+        Data to cluster
+
+    n_clusters: int,
+        The number of clusters to find.
+
+    init : {'k-means++', 'random' or an ndarray}
+        Method for initialization, defaults to 'k-means++':
+        'k-means++' : selects initial cluster centers for k-means
+        clustering in a smart way to speed up convergence. See section
+        Notes in k_init for more details.
+        'random': choose k observations (rows) at random from data for
+        the initial centroids.
+        If an ndarray is passed, it should be of shape (n_clusters, n_features)
+        and gives the initial centers.
+
+    batch_size : int, optional, default: 1000
+        Size of the mini batches. (Kmeans performed through MiniBatchKMeans)
+
+    n_init : int, default=10
+        Number of random initializations that are tried.
+        In contrast to KMeans, the algorithm is only run once, using the
+        best of the ``n_init`` initializations as measured by inertia.
+
+    max_no_improvement : int, default: 10
+        Control early stopping based on the consecutive number of mini
+        batches that does not yield an improvement on the smoothed inertia.
+        To disable convergence detection based on inertia, set
+        max_no_improvement to None.
+
+    random_state : int, RandomState instance or None (default)
+        Determines random number generation for centroid initialization and
+        random reassignment. Use an int to make the randomness deterministic.
+
+    Returns
+    -------
+    labels : list of ints (len n_features)
+        Parcellation of features in clusters
+    """
+
+    n_big_clusters = int(np.sqrt(n_clusters))
+    mbk = MiniBatchKMeans(init=init, n_clusters=n_big_clusters,
+                          batch_size=batch_size, n_init=n_init,
+                          max_no_improvement=max_no_improvement,
+                          verbose=verbose, random_state=random_state).fit(X)
+    coarse_labels = mbk.labels_
+    fine_labels = np.zeros_like(coarse_labels)
+    q = 0
+    exact_clusters = np.asarray([n_clusters * np.sum(coarse_labels == i) * 1.
+                                 / X.shape[0] for i in range(n_big_clusters)])
+
+    adjusted_clusters = _adjust_small_clusters(exact_clusters, n_clusters)
+    for i, n_small_clusters in enumerate(adjusted_clusters):
+        mbk = MiniBatchKMeans(init=init, n_clusters=n_small_clusters,
+                              batch_size=batch_size, random_state=random_state,
+                              max_no_improvement=max_no_improvement,
+                              verbose=verbose,
+                              n_init=n_init,).fit(X[coarse_labels == i])
+        fine_labels[coarse_labels == i] = q + mbk.labels_
+        q += n_small_clusters
+
+    return _remove_empty_labels(fine_labels)
```

Comment:
would it make sense to also return the coarse labels if someone wants to see the hierarchy?